projects/output-and-error-handling/client/main.go

```diff
@@ -0,0 +1,92 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+func main() {
+	retries := 0
+	// This loop will run only if the number of reties doesn't exceed 3
+	for retries < 3 {
+		resp := getWeather()
+		// Close the response body after been fully read
+		defer resp.Body.Close()
+
+		// Read response body and store it in a var
+		body, err := io.ReadAll(resp.Body)
+		// Show error message if we cannot read the response body
+		if err != nil {
+			fmt.Print("Sorry we cannot get the weather!\n")
+			fmt.Fprintf(os.Stderr, "Failed to read response body: %v\n", err)
+			os.Exit(1)
+		}
+
+		// Handle cases depending on the Status code of the response
+		switch resp.StatusCode {
+		case 200:
+			// Set the retries number to 0 so we can run the program many time when the response is 200
+			retries = 0
+
+			// Convert response body from binary to string
+			sb := string(body)
+			fmt.Fprint(os.Stdout, sb+"\n")
+			os.Exit(0)
+		case 429:
+			handleRateLimited(resp.Header.Get("Retry-After"), retries)
+		case 500:
+			fmt.Print("Sorry we cannot get the weather!\n")
+			fmt.Fprint(os.Stderr, "Internal Server Error\n")
+			os.Exit(1)
+		default:
+			fmt.Print("Sorry we cannot get the weather!\n")
+			fmt.Fprint(os.Stderr, "Unexpected Error\n")
+			os.Exit(1)
+		}
+	}
+}
+
+func getWeather() *http.Response {
+	// Connect to the server and getting a response
+	resp, err := http.Get("http://localhost:8080")
+	// Show error message if connection is not established
+	if err != nil {
+		fmt.Print("Sorry we cannot get the weather!\n")
+		fmt.Fprintf(os.Stderr, "Failed to make http request: %v\n", err)
+		os.Exit(1)
+	}
+	return resp
+}
+
+// Handle response and retry depending on the Retry-After header
+func handleRateLimited(retryTime string, retries int) {
+	retrySeconds := 0
+	var err error
+	retryTimeDate, err := time.Parse(time.RFC1123, retryTime)
+
+	if err == nil {
+		retrySeconds = int(time.Until(retryTimeDate).Seconds())
+	} else if retryTime == "a while" {
+		retrySeconds = 5
+	} else {
+		retrySeconds, err = strconv.Atoi(retryTime)
+		if err != nil {
+			fmt.Print("Sorry we cannot get the weather!")
+			fmt.Fprintf(os.Stderr, "Internal Error : Failed to convert retry time: %v\n", err)
+			os.Exit(1)
+		}
+	}
+	if retrySeconds > 1 && retrySeconds <= 5 {
+		fmt.Printf("We will retry to get you the weather. Please wait %d seconds\n", retrySeconds)
+		time.Sleep(time.Duration(retrySeconds) * time.Second)
```

Comment:
This sleep is going to make it quite slow to run any tests you may write about parsing this header - can you think of a way you could refactor the code so you can _quickly_ run tests for your parsing logic?

Reply:
I would think about separating the "work out how long we need to sleep" function from the "actually sleep" part - if you introduced a separate function for the `string` -> `time.Duration` calculation which you call in this function, you can test that without actually needing to sleep.